// @shoelace-style/shoelace/dist/all.shoelace.js
import "./chunk.3I42H3S6.js";
import {
  textarea_default
} from "./chunk.SQHEG5PX.js";
import {
  tooltip_default
} from "./chunk.GVM3FDPO.js";
import {
  select_default
} from "./chunk.4HEACYII.js";
import {
  skeleton_default
} from "./chunk.QDN2XRN7.js";
import {
  spinner_default
} from "./chunk.DKNVMYJL.js";
import {
  switch_default
} from "./chunk.ZP6DW5ES.js";
import {
  tab_default
} from "./chunk.54GSUZ3E.js";
import {
  tab_group_default
} from "./chunk.XCFNNUZB.js";
import {
  tab_panel_default
} from "./chunk.QRWXHFX7.js";
import {
  tag_default
} from "./chunk.XV3DXFBP.js";
import {
  progress_bar_default
} from "./chunk.UX3DH266.js";
import {
  progress_ring_default
} from "./chunk.B54RLEFV.js";
import {
  radio_default
} from "./chunk.YUPRDP44.js";
import {
  range_default
} from "./chunk.3PUUGC7M.js";
import {
  rating_default
} from "./chunk.4OUZEE2Z.js";
import {
  relative_time_default
} from "./chunk.2FYMRJKS.js";
import {
  resize_observer_default
} from "./chunk.UUNPRXKU.js";
import {
  responsive_embed_default
} from "./chunk.SWHWG4QP.js";
import {
  image_comparer_default
} from "./chunk.CYIXBC7Q.js";
import {
  include_default
} from "./chunk.F3R7OMVC.js";
import "./chunk.2HYO2JO7.js";
import {
  input_default
} from "./chunk.YJAHKJ77.js";
import "./chunk.HISVFV44.js";
import {
  menu_default
} from "./chunk.ZEQTQWMJ.js";
import "./chunk.U2XCTTGK.js";
import {
  menu_divider_default
} from "./chunk.L4M2MKU5.js";
import {
  menu_item_default
} from "./chunk.67CH4Y26.js";
import {
  menu_label_default
} from "./chunk.57OE76OH.js";
import {
  form_default
} from "./chunk.WALK6O7W.js";
import {
  format_bytes_default
} from "./chunk.VNSVRZWC.js";
import {
  format_date_default
} from "./chunk.B3SIPET3.js";
import {
  format_number_default
} from "./chunk.CXH4ABOX.js";
import {
  icon_default
} from "./chunk.IG443PQN.js";
import {
  registerIconLibrary,
  unregisterIconLibrary
} from "./chunk.AF62ZLUK.js";
import {
  getBasePath,
  setBasePath
} from "./chunk.YALGRZIW.js";
import "./chunk.7GTYP37T.js";
import {
  icon_button_default
} from "./chunk.NGLAU5ZY.js";
import {
  button_group_default
} from "./chunk.3I3MUR43.js";
import {
  card_default
} from "./chunk.PBYBDRLF.js";
import {
  checkbox_default
} from "./chunk.Q45P5IPR.js";
import {
  color_picker_default
} from "./chunk.SK35BAJ5.js";
import "./chunk.S6SAZ7RG.js";
import "./chunk.IA2H52LP.js";
import {
  details_default
} from "./chunk.RXBE3UPW.js";
import "./chunk.GQSSQUA6.js";
import {
  dialog_default
} from "./chunk.7CVLJZNV.js";
import {
  drawer_default
} from "./chunk.XRNA4NHC.js";
import "./chunk.5LDML7H2.js";
import {
  dropdown_default
} from "./chunk.L64C37GJ.js";
import "./chunk.POBXNCHG.js";
import "./chunk.LDX6FEXO.js";
import "./chunk.TKRDPH6K.js";
import {
  alert_default
} from "./chunk.4D36RQEL.js";
import {
  animation_default
} from "./chunk.GYQRGKMO.js";
import {
  getAnimationNames,
  getEasingNames
} from "./chunk.FFLIIME4.js";
import {
  __export
} from "./chunk.FBS5DUDL.js";
import "./chunk.W2OJMTVT.js";
import {
  avatar_default
} from "./chunk.46EZWOEK.js";
import {
  badge_default
} from "./chunk.5BABZGQL.js";
import {
  button_default
} from "./chunk.WSHTIYZN.js";
import "./chunk.PCLPGKVT.js";
import "./chunk.IJFXZJ6H.js";
import "./chunk.HODVHBF3.js";

// src/shoelace.ts
var shoelace_exports = {};
__export(shoelace_exports, {
  SlAlert: () => alert_default,
  SlAnimation: () => animation_default,
  SlAvatar: () => avatar_default,
  SlBadge: () => badge_default,
  SlButton: () => button_default,
  SlButtonGroup: () => button_group_default,
  SlCard: () => card_default,
  SlCheckbox: () => checkbox_default,
  SlColorPicker: () => color_picker_default,
  SlDetails: () => details_default,
  SlDialog: () => dialog_default,
  SlDrawer: () => drawer_default,
  SlDropdown: () => dropdown_default,
  SlForm: () => form_default,
  SlFormatBytes: () => format_bytes_default,
  SlFormatDate: () => format_date_default,
  SlFormatNumber: () => format_number_default,
  SlIcon: () => icon_default,
  SlIconButton: () => icon_button_default,
  SlImageComparer: () => image_comparer_default,
  SlInclude: () => include_default,
  SlInput: () => input_default,
  SlMenu: () => menu_default,
  SlMenuDivider: () => menu_divider_default,
  SlMenuItem: () => menu_item_default,
  SlMenuLabel: () => menu_label_default,
  SlProgressBar: () => progress_bar_default,
  SlProgressRing: () => progress_ring_default,
  SlRadio: () => radio_default,
  SlRange: () => range_default,
  SlRating: () => rating_default,
  SlRelativeTime: () => relative_time_default,
  SlResizeObserver: () => resize_observer_default,
  SlResponsiveEmbed: () => responsive_embed_default,
  SlSelect: () => select_default,
  SlSkeleton: () => skeleton_default,
  SlSpinner: () => spinner_default,
  SlSwitch: () => switch_default,
  SlTab: () => tab_default,
  SlTabGroup: () => tab_group_default,
  SlTabPanel: () => tab_panel_default,
  SlTag: () => tag_default,
  SlTextarea: () => textarea_default,
  SlTooltip: () => tooltip_default,
  getAnimationNames: () => getAnimationNames,
  getBasePath: () => getBasePath,
  getEasingNames: () => getEasingNames,
  registerIconLibrary: () => registerIconLibrary,
  setBasePath: () => setBasePath,
  unregisterIconLibrary: () => unregisterIconLibrary
});

// src/all.shoelace.ts
Object.keys(shoelace_exports).map((key) => {
  const item = shoelace_exports[key];
  if (typeof item.register === "function") {
    item.register();
  }
});
export {
  alert_default as SlAlert,
  animation_default as SlAnimation,
  avatar_default as SlAvatar,
  badge_default as SlBadge,
  button_default as SlButton,
  button_group_default as SlButtonGroup,
  card_default as SlCard,
  checkbox_default as SlCheckbox,
  color_picker_default as SlColorPicker,
  details_default as SlDetails,
  dialog_default as SlDialog,
  drawer_default as SlDrawer,
  dropdown_default as SlDropdown,
  form_default as SlForm,
  format_bytes_default as SlFormatBytes,
  format_date_default as SlFormatDate,
  format_number_default as SlFormatNumber,
  icon_default as SlIcon,
  icon_button_default as SlIconButton,
  image_comparer_default as SlImageComparer,
  include_default as SlInclude,
  input_default as SlInput,
  menu_default as SlMenu,
  menu_divider_default as SlMenuDivider,
  menu_item_default as SlMenuItem,
  menu_label_default as SlMenuLabel,
  progress_bar_default as SlProgressBar,
  progress_ring_default as SlProgressRing,
  radio_default as SlRadio,
  range_default as SlRange,
  rating_default as SlRating,
  relative_time_default as SlRelativeTime,
  resize_observer_default as SlResizeObserver,
  responsive_embed_default as SlResponsiveEmbed,
  select_default as SlSelect,
  skeleton_default as SlSkeleton,
  spinner_default as SlSpinner,
  switch_default as SlSwitch,
  tab_default as SlTab,
  tab_group_default as SlTabGroup,
  tab_panel_default as SlTabPanel,
  tag_default as SlTag,
  textarea_default as SlTextarea,
  tooltip_default as SlTooltip,
  getAnimationNames,
  getBasePath,
  getEasingNames,
  registerIconLibrary,
  setBasePath,
  unregisterIconLibrary
};
