// @shoelace-style/shoelace/dist/chunk.F3R7OMVC.js
import {
  requestInclude
} from "./chunk.2HYO2JO7.js";
import {
  Shoemaker,
  html
} from "./chunk.HODVHBF3.js";

// _ccwmwdlsd:/Users/claviska/Projects/shoelace/src/components/include/include.scss
var include_default = ":host {\n  display: block;\n}";

// src/components/include/include.ts
var SlInclude = class extends Shoemaker {
  constructor() {
    super(...arguments);
    this.mode = "cors";
    this.allowScripts = false;
  }
  watchSrc() {
    this.loadSource();
  }
  onConnect() {
    this.loadSource();
  }
  executeScript(script) {
    const newScript = document.createElement("script");
    [...script.attributes].map((attr) => newScript.setAttribute(attr.name, attr.value));
    newScript.textContent = script.textContent;
    script.parentNode.replaceChild(newScript, script);
  }
  async loadSource() {
    try {
      const src = this.src;
      const file = await requestInclude(src, this.mode);
      if (src !== this.src) {
        return;
      }
      if (!file) {
        return;
      }
      if (!file.ok) {
        this.emit("sl-error", {detail: {status: file.status}});
        return;
      }
      this.innerHTML = file.html;
      if (this.allowScripts) {
        [...this.querySelectorAll("script")].map((script) => this.executeScript(script));
      }
      this.emit("sl-load");
    } catch (e) {
      this.emit("sl-error");
    }
  }
  render() {
    return html`<slot />`;
  }
};
SlInclude.tag = "sl-include";
SlInclude.props = ["html", "src", "mode", " allowScripts"];
SlInclude.styles = include_default;
var include_default2 = SlInclude;

export {
  include_default2 as include_default
};
