// @shoelace-style/shoelace/dist/chunk.CXH4ABOX.js
import {
  Shoemaker
} from "./chunk.HODVHBF3.js";

// src/components/format-number/format-number.ts
var SlFormatNumber = class extends Shoemaker {
  constructor() {
    super(...arguments);
    this.value = 0;
    this.type = "decimal";
    this.noGrouping = false;
    this.currency = "USD";
    this.currencyDisplay = "symbol";
  }
  render() {
    if (isNaN(this.value)) {
      return "";
    }
    return new Intl.NumberFormat(this.locale, {
      style: this.type,
      currency: this.currency,
      currencyDisplay: this.currencyDisplay,
      useGrouping: !this.noGrouping,
      minimumIntegerDigits: this.minimumIntegerDigits,
      minimumFractionDigits: this.minimumFractionDigits,
      maximumFractionDigits: this.maximumFractionDigits,
      minimumSignificantDigits: this.minimumSignificantDigits,
      maximumSignificantDigits: this.maximumSignificantDigits
    }).format(this.value);
  }
};
SlFormatNumber.tag = "sl-format-number";
SlFormatNumber.props = [
  "value",
  "locale",
  "type",
  "noGrouping",
  "currency",
  "currencyDisplay",
  "minimumIntegerDigits",
  "minimumFractionDigits",
  "maximumFractionDigits",
  "minimumSignificantDigits",
  "maximumSignificantDigits"
];
var format_number_default = SlFormatNumber;

export {
  format_number_default
};
