// @shoelace-style/shoelace/dist/chunk.NGLAU5ZY.js
import {
  focusVisible
} from "./chunk.GQSSQUA6.js";
import {
  classMap
} from "./chunk.IJFXZJ6H.js";
import {
  Shoemaker,
  html
} from "./chunk.HODVHBF3.js";

// _ccwmwdlsd:/Users/claviska/Projects/shoelace/src/components/icon-button/icon-button.scss
var icon_button_default = ":host {\n  position: relative;\n  box-sizing: border-box;\n}\n:host *, :host *:before, :host *:after {\n  box-sizing: inherit;\n}\n\n:host {\n  display: inline-block;\n}\n\n.icon-button {\n  flex: 0 0 auto;\n  display: flex;\n  align-items: center;\n  background: none;\n  border: none;\n  border-radius: var(--sl-border-radius-medium);\n  font-size: inherit;\n  color: var(--sl-color-gray-500);\n  padding: var(--sl-spacing-x-small);\n  cursor: pointer;\n  transition: var(--sl-transition-medium) color;\n  -webkit-appearance: none;\n}\n.icon-button:hover:not(.icon-button--disabled), .icon-button:focus:not(.icon-button--disabled) {\n  color: var(--sl-color-primary-500);\n}\n.icon-button:active:not(.icon-button--disabled) {\n  color: var(--sl-color-primary-600);\n}\n.icon-button:focus {\n  outline: none;\n}\n\n.icon-button--disabled {\n  opacity: 0.5;\n  cursor: not-allowed;\n}\n\n.focus-visible.icon-button:focus {\n  box-shadow: 0 0 0 var(--sl-focus-ring-width) var(--sl-focus-ring-color-primary);\n}";

// src/components/icon-button/icon-button.ts
var SlIconButton = class extends Shoemaker {
  constructor() {
    super(...arguments);
    this.disabled = false;
  }
  onReady() {
    focusVisible.observe(this.button);
  }
  disconnectedCallback() {
    focusVisible.unobserve(this.button);
  }
  render() {
    return html`
      <button
        ref=${(el) => this.button = el}
        part="base"
        class=${classMap({
      "icon-button": true,
      "icon-button--disabled": this.disabled
    })}
        type="button"
        aria-label=${this.label}
      >
        <sl-icon library=${this.library} name=${this.name} src=${this.src} aria-hidden="true" />
      </button>
    `;
  }
};
SlIconButton.tag = "sl-icon-button";
SlIconButton.props = ["name", "library", "src", "label", "disabled"];
SlIconButton.reflect = ["disabled"];
SlIconButton.styles = icon_button_default;
var icon_button_default2 = SlIconButton;

export {
  icon_button_default2 as icon_button_default
};
