// @shoelace-style/shoelace/dist/chunk.SWHWG4QP.js
import {
  Shoemaker,
  html
} from "./chunk.HODVHBF3.js";

// _ccwmwdlsd:/Users/claviska/Projects/shoelace/src/components/responsive-embed/responsive-embed.scss
var responsive_embed_default = ":host {\n  position: relative;\n  box-sizing: border-box;\n}\n:host *, :host *:before, :host *:after {\n  box-sizing: inherit;\n}\n\n:host {\n  display: block;\n}\n\n.responsive-embed {\n  position: relative;\n}\n.responsive-embed ::slotted(embed),\n.responsive-embed ::slotted(iframe),\n.responsive-embed ::slotted(object) {\n  position: absolute !important;\n  top: 0 !important;\n  left: 0 !important;\n  width: 100% !important;\n  height: 100% !important;\n}";

// src/components/responsive-embed/responsive-embed.ts
var SlResponsiveEmbed = class extends Shoemaker {
  constructor() {
    super(...arguments);
    this.aspectRatio = "16:9";
  }
  updateAspectRatio() {
    const split = this.aspectRatio.split(":");
    const x = parseInt(split[0]);
    const y = parseInt(split[1]);
    this.base.style.paddingBottom = x && y ? `${y / x * 100}%` : "";
  }
  watchAspectRatio() {
    this.updateAspectRatio();
  }
  render() {
    return html`
      <div ref=${(el) => this.base = el} part="base" class="responsive-embed">
        <slot onslotchange=${() => this.updateAspectRatio()} />
      </div>
    `;
  }
};
SlResponsiveEmbed.tag = "sl-responsive-embed";
SlResponsiveEmbed.props = ["aspectRatio"];
SlResponsiveEmbed.styles = responsive_embed_default;
var responsive_embed_default2 = SlResponsiveEmbed;

export {
  responsive_embed_default2 as responsive_embed_default
};
